tests/functional/s3/test_sync_command.py

```diff
@@ -164,3 +165,17 @@ def test_sync_with_delete_on_downloads(self):
         self.assertEqual(self.operations_called[0][0].name, 'ListObjects')
 
         self.assertFalse(os.path.exists(full_path))
+
+    def test_sync_skips_over_files_deleted_between_listing_and_transfer_attempt(self):
+        full_path = self.files.create_file('foo.txt', 'mycontent')
+        cmdline = '%s %s s3://bucket/' % (
+            self.prefix, self.files.rootdir)
+
+        # FileGenerator.list_files should skip over files that cause an OSError to be raised
+        # because they are missing when we try to get their stats.
+        with patch('awscli.customizations.s3.filegenerator.get_file_stat', side_effect=OSError()):
+            self.run_cmd(cmdline, expected_rc=0)
```

Comment:
I agree, it's an improvement to sync as much as possible. The problem in https://github.com/aws/aws-cli/issues/1841 is that we want to ignore these errors - that is, we need at least need a way to reliably detect this particular error scenario so we can let our scripts consider the sync a success, if not have the `s3 sync` command itself exit with 0 – and I am not sure if I will be able to say I resolved that issue with this PR if the command exits with 2, because I don't know if there is a reliable way to detect this particular error scenario.

I have not tried this patch on my current setup. I'm not able to reliably reproduce the issue, so unless it's necessary to test it, I'd like to avoid the pain of doing so.